File: services/annotation/app/schemas/train_agent.py
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field

from app.schemas.train_task import YoloTrainConfig


class AgentChatRequest(BaseModel):
    session_id: Optional[str] = None
    task_name: str = Field(min_length=1, max_length=255)
    message: str = Field(min_length=1, max_length=10000)
    current_config: Optional[Dict[str, Any]] = None
    train_task_id: Optional[int] = None

    class Config:
        extra = "forbid"


class ConfigChange(BaseModel):
    field: str
    before: Any = None
    after: Any = None
    reason: str


class AgentModelResult(BaseModel):
    reply: str = Field(min_length=1)
    config_patch: Dict[str, Any] = Field(default_factory=dict)
    changes: List[ConfigChange] = Field(default_factory=list)
    questions: List[str] = Field(default_factory=list)
    warnings: List[str] = Field(default_factory=list)
    ready_to_apply: bool = False

    class Config:
        extra = "forbid"


class AgentChatData(BaseModel):
    session_id: str
    mode: str
    reply: str
    proposal_id: Optional[str] = None
    config: Optional[YoloTrainConfig] = None
    changes: List[ConfigChange] = Field(default_factory=list)
    questions: List[str] = Field(default_factory=list)
    warnings: List[str] = Field(default_factory=list)
    ready_to_apply: bool = False
    analysis: Optional[Dict[str, Any]] = None


class ConfirmProposalRequest(BaseModel):
    expected_config: Dict[str, Any]

    class Config:
        extra = "forbid"


class StartProposalRequest(BaseModel):
    confirmation_token: str = Field(min_length=32, max_length=512)

    class Config:
        extra = "forbid"


class CreateOptimizationProposalRequest(BaseModel):
    instruction: str = Field(
        default="基于本次训练报告生成下一轮优化训练参数草案",
        min_length=1,
        max_length=5000,
    )

    class Config:
        extra = "forbid"

File: apps/web/src/types/trainAgent.ts
export type TrainAgentConfig = Record<string, unknown>

export interface TrainOptimizationDraft {
  config: TrainAgentConfig
  sourceTrainTaskId: number
}

export interface TrainAgentChatRequest {
  session_id: string | null
  task_name: string
  message: string
  current_config: TrainAgentConfig
  train_task_id: number | null
}

export interface TrainAgentChatResponse {
  session_id?: string | null
  proposal_id?: string
  config?: TrainAgentConfig
  message?: string
  reply?: string
  content?: string
  changes?: Array<{
    field: string
    before?: unknown
    after?: unknown
    reason: string
  }>
  warnings?: string[]
  questions?: string[]
  ready_to_apply?: boolean
  [key: string]: unknown
}

export interface TrainAgentContext {
  image_count?: number
  total_images?: number
  annotation_count?: number
  total_annotations?: number
  category_distribution?: unknown
  object_size_distribution?: unknown
  available_models?: unknown
  [key: string]: unknown
}

export interface TrainAgentConfirmResponse {
  confirmation_token?: string
  expires_at?: string
  expires_in?: number
  [key: string]: unknown
}

export interface TrainAgentStartResponse {
  proposal_id?: string
  train_task_id?: number
  status?: string
  task_id?: number
  id?: number
  [key: string]: unknown
}

export type TrainAgentAnalysis = Record<string, unknown>

export type TrainAgentAutoAnalysisStatus = 'PENDING' | 'RUNNING' | 'COMPLETED' | 'FAILED'

export interface TrainAgentAutoAnalysis {
  train_task_id?: number
  status: TrainAgentAutoAnalysisStatus
  analysis?: TrainAgentAnalysis | null
  model_result?: {
    reply?: string
    [key: string]: unknown
  } | null
  error_message?: string | null
  [key: string]: unknown
}
